client/components/domains/registrant-extra-info/fr-validate-contact-details.js

```diff
@@ -74,9 +74,17 @@ export default function validateContactDetails( contactDetails ) {
 				.split( '.' )
 				.slice( 1 );
 
+			// In order to capture the relationship between the organization
+			// and extra.individualType fields, the rule ends up in the root
+			// path.
+			// We've only got one such case at the moment, so we can insert this
+			// hack, but if we need to tell multiple such rules apart, we're
+			// going to need to add a some magic to map schemas to fields
```

Comment:
How hard would said map be? I don't understand this part of the code well but hack sense is tingling and if the investment is straightward and cheap enough (say, half a day?) my vote would be worthwhile.

If it'd be a big deal, then, ignore this.

Reply:
Yeah, "hack sense is tingling" is spot on.

The map itself would easy, but it would just be another hack. Json schema can totally do this, so we'd be starting down the slippery slope of writing our own schema validator, and closing ourselves off from the spec.

The root of the problem is that we don't get good information about the failing rule. If we could grab the rule that failed, we can add any meta data that we want.

For example, in this case we want to be able to specify that this rule describes a problem with the organization field, so we could use a convention where a `field` attribute overrides the path if it's present and add `field: 'organization'` to the rule in the schema to say "when this rule fails, it indicates a problem with the organization field. We could also add an `errorCode`, or even `errorMessage` directly, and maybe also the variable extraction that @klimeryk was musing about to feed variables into error messages (e.g. naming the specific banned word in the error message). In short: whatever we need.

`is-my-json-valid` is excellent in terms of getting pass/fail right, it's just a bit terse. There are several issues related to it that have been open for several years (https://github.com/mafintosh/is-my-json-valid/issues/38, https://github.com/mafintosh/is-my-json-valid/issues/39, https://github.com/mafintosh/is-my-json-valid/issues/22).

Trying to fix that looks pretty fun but not very easy, so at the moment I'm leaning towards trying a different lib ([ajv](https://github.com/epoberezkin/ajv) in particular looks very promising), and that requires discussion and investigation that I'd prefer to do after landing this bugfix.

Reply:
Mini essay in comments, short version: let's not in this PR.